File: cpp_version/src/network/tcp_server.cpp
#include "network/tcp_server.h"
#include <sys/socket.h>
#include <sys/epoll.h>
#include <netinet/in.h>
#include <unistd.h>
#include <fcntl.h>
#include <stdexcept>
#include <cstring>
#include <iostream>

TcpServer::TcpServer(Reactor& reactor, const std::string& ip, int port)
    : reactor_(reactor), ip_(ip), port_(port), running_(false) {
    server_fd_ = socket(AF_INET, SOCK_STREAM, 0);
    if (server_fd_ == -1) {
        throw std::runtime_error("Failed to create socket");
    }

    int opt = 1;
    if (setsockopt(server_fd_, SOL_SOCKET, SO_REUSEADDR, &opt, sizeof(opt)) == -1) {
        close(server_fd_);
        throw std::runtime_error("Failed to set socket options");
    }

    sockaddr_in address;
    address.sin_family = AF_INET; // 设置地址族为 IPv4（AF_INET）。
    address.sin_addr.s_addr = INADDR_ANY;  // 这行代码是用来设置服务器监听的 IP 地址，INADDR_ANY表示服务器将在所有可用的网络接口上监听。
    address.sin_port = htons(port_); // 将端口号转换为网络字节序。

    if (bind(server_fd_, (struct sockaddr*)&address, sizeof(address)) == -1) {
        close(server_fd_);
        throw std::runtime_error("Failed to bind to port");
    }

    if (listen(server_fd_, SOMAXCONN) == -1) {
        close(server_fd_);
        throw std::runtime_error("Failed to listen on socket");
    }

    // 把我们的服务端设置为非阻塞模式
    int flags = fcntl(server_fd_, F_GETFL, 0);
    fcntl(server_fd_, F_SETFL, flags | O_NONBLOCK);
}

TcpServer::~TcpServer() {
    if (running_) {
        stop();
    }
    close(server_fd_);
}

void TcpServer::start() {
    running_ = true;
    reactor_.add_handler(server_fd_, EPOLLIN, [this](uint32_t events) {
        accept_connection();
    });
}

void TcpServer::stop() {
    running_ = false;
    reactor_.remove_handler(server_fd_);
}

void TcpServer::set_receive_handler(std::function<void(int, const char*, size_t)> handler) {
    receive_handler_ = std::move(handler);
}

void TcpServer::send(int client_fd, const char* data, size_t len) {
    ::send(client_fd, data, len, 0);
}

void TcpServer::handle_close(int client_fd) {
    reactor_.remove_handler(client_fd);
    close(client_fd);
}

void TcpServer::accept_connection() {
    sockaddr_in client_addr;
    socklen_t client_len = sizeof(client_addr);
    int client_fd = accept(server_fd_, (struct sockaddr*)&client_addr, &client_len);

    if (client_fd == -1) {
        if (errno != EAGAIN && errno != EWOULDBLOCK) {
            std::cerr << "Failed to accept connection" << std::endl;
        }
        return;
    }

    // 为客户端的socket设置非阻塞
    int flags = fcntl(client_fd, F_GETFL, 0);
    fcntl(client_fd, F_SETFL, flags | O_NONBLOCK);

    reactor_.add_handler(client_fd, EPOLLIN | EPOLLRDHUP, [this, client_fd](uint32_t events) {
        if (events & EPOLLIN) {
            handle_read(client_fd);
        }
        if (events & (EPOLLRDHUP | EPOLLHUP)) {
            handle_close(client_fd);
        }
    });
}

void TcpServer::handle_read(int client_fd) {
    char buffer[BUFFER_SIZE];
    ssize_t bytes_read = read(client_fd, buffer, BUFFER_SIZE);

    if (bytes_read > 0) {
        if (receive_handler_) {
            receive_handler_(client_fd, buffer, bytes_read);
        }
    } else if (bytes_read == 0 || (bytes_read == -1 && errno != EAGAIN)) {
        handle_close(client_fd);
    }
}

File: rust_version/src/network/tcp_server.rs
use crate::core::reactor::Reactor;
use std::io;
use std::os::unix::io::{AsRawFd, RawFd, IntoRawFd};
use std::net::TcpListener;
use std::sync::{Arc, Mutex};
use std::sync::atomic::{AtomicBool, Ordering};

const BUFFER_SIZE: usize = 1024;

struct ServerState {
    reactor: Reactor,
    receive_handler: Option<Box<dyn FnMut(RawFd, &[u8], usize) + Send>>,
}

#[allow(dead_code)]
pub struct TcpServer {
    state: Arc<Mutex<ServerState>>,
    ip: String,
    port: u16,
    server_fd: RawFd,
    running: Arc<AtomicBool>,
}

impl TcpServer {
    pub fn new(reactor: Reactor, ip: &str, port: u16) -> io::Result<Self> {
        let addr = format!("{}:{}", ip, port);
        let listener = TcpListener::bind(&addr)?;
        
        listener.set_nonblocking(true)?;
        let listen_fd = listener.as_raw_fd();
        
        unsafe {
            let opt: libc::c_int = 1;
            if libc::setsockopt(
                listen_fd,
                libc::SOL_SOCKET,
                libc::SO_REUSEADDR,
                &opt as *const _ as *const libc::c_void,
                std::mem::size_of_val(&opt) as libc::socklen_t,
            ) == -1 {
                return Err(io::Error::last_os_error());
            }
        }
        
        let _ = listener.into_raw_fd();
        
        Ok(TcpServer {
            state: Arc::new(Mutex::new(ServerState {
                reactor,
                receive_handler: None,
            })),
            ip: ip.to_string(),
            port,
            server_fd: listen_fd,
            running: Arc::new(AtomicBool::new(false)),
        })
    }

    fn handle_read(client_fd: RawFd, state: &Arc<Mutex<ServerState>>) -> io::Result<()> {
        let mut buffer = [0u8; BUFFER_SIZE];
        
        unsafe {
            let bytes_read = libc::read(
                client_fd,
                buffer.as_mut_ptr() as *mut libc::c_void,
                buffer.len(),
            );
            
            if bytes_read > 0 {
                if let Some(ref mut handler) = state.lock().unwrap().receive_handler {
                    handler(client_fd, &buffer[..bytes_read as usize], bytes_read as usize);
                }
            } else if bytes_read == 0 || (bytes_read == -1 && io::Error::last_os_error().kind() != io::ErrorKind::WouldBlock) {
                Self::handle_close(client_fd, state)?;
            }
        }
        Ok(())
    }

    fn handle_close(client_fd: RawFd, state: &Arc<Mutex<ServerState>>) -> io::Result<()> {
        state.lock().unwrap().reactor.remove_handler(client_fd)?;
        unsafe {
            if libc::close(client_fd) == -1 {
                return Err(io::Error::last_os_error());
            }
        }
        Ok(())
    }

    pub fn accept_connection(&mut self) -> io::Result<()> {
        unsafe {
            let mut client_addr: libc::sockaddr_in = std::mem::zeroed();
            let mut client_len = std::mem::size_of::<libc::sockaddr_in>() as libc::socklen_t;
            
            let client_fd = libc::accept(
                self.server_fd,
                &mut client_addr as *mut _ as *mut libc::sockaddr,
                &mut client_len,
            );
            
            if client_fd == -1 {
                let err = io::Error::last_os_error();
                if err.kind() != io::ErrorKind::WouldBlock {
                    eprintln!("Failed to accept connection: {}", err);
                }
                return Ok(());
            }
            
            let flags = libc::fcntl(client_fd, libc::F_GETFL, 0);
            if flags == -1 {
                return Err(io::Error::last_os_error());
            }
            if libc::fcntl(client_fd, libc::F_SETFL, flags | libc::O_NONBLOCK) == -1 {
                return Err(io::Error::last_os_error());
            }

            let state = Arc::clone(&self.state);
            self.state.lock().unwrap().reactor.add_handler(
                client_fd,
                (libc::EPOLLIN | libc::EPOLLRDHUP) as u32,
                Box::new(move |events| {
                    if events & (libc::EPOLLIN as u32) != 0 {
                        if let Err(e) = Self::handle_read(client_fd, &state) {
                            eprintln!("Error handling read: {}", e);
                        }
                    }
                    if events & ((libc::EPOLLRDHUP | libc::EPOLLHUP) as u32) != 0 {
                        if let Err(e) = Self::handle_close(client_fd, &state) {
                            eprintln!("Error closing connection: {}", e);
                        }
                    }
                }),
            )?;
        }
        Ok(())
    }

    pub fn start(&mut self) -> io::Result<()> {
        self.running.store(true, Ordering::SeqCst);
        let state = Arc::clone(&self.state);
        let server_fd = self.server_fd;
        let running = Arc::clone(&self.running);

        self.state.lock().unwrap().reactor.add_handler(
            self.server_fd,
            libc::EPOLLIN as u32,
            Box::new(move |events| {
                if events & (libc::EPOLLIN as u32) != 0 {
                    let mut server = TcpServer {
                        state: Arc::clone(&state),
                        ip: String::new(),
                        port: 0,
                        server_fd,
                        running: running.clone(),
                    };
                    if let Err(e) = server.accept_connection() {
                        eprintln!("Accept error happened: {}", e);
                    }
                }
            }),
        )?;
        
        Ok(())
    }

    pub fn stop(&mut self) -> io::Result<()> {
        println!("TcpServer stopping...");
        self.running.store(false, Ordering::SeqCst);
        self.state.lock().unwrap().reactor.remove_handler(self.server_fd)?;
        self.state.lock().unwrap().reactor.stop();
        println!("TcpServer stopped");
        Ok(())
    }

    pub fn set_receive_handler<F>(&mut self, handler: F)
    where
        F: FnMut(RawFd, &[u8], usize) + Send + 'static,
    {
        self.state.lock().unwrap().receive_handler = Some(Box::new(handler));
    }

    pub fn send(&self, client_fd: RawFd, data: &[u8]) -> io::Result<usize> {
        unsafe {
            let sent = libc::send(
                client_fd,
                data.as_ptr() as *const libc::c_void,
                data.len(),
                0,
            );
            if sent == -1 {
                return Err(io::Error::last_os_error());
            }
            Ok(sent as usize)
        }
    }

    pub fn get_reactor(&self) -> Reactor {
        let mut reactor = self.state.lock().unwrap().reactor.clone();
        reactor.share_running_state(Arc::clone(&self.running));
        reactor
    }
}

impl Drop for TcpServer {
    fn drop(&mut self) {
        if self.running.load(Ordering::SeqCst) {
            let _ = self.stop();
        }
        unsafe {
            let _ = libc::close(self.server_fd);
        }
    }
}